admin/server/auth/claims.go

```diff
@@ -75,3 +87,144 @@ func (c *authTokenClaims) OwnerID() string {
 func (c *authTokenClaims) AuthTokenID() string {
 	return c.token.Token().ID.String()
 }
+
+func (c *authTokenClaims) CanOrganization(ctx context.Context, orgID string, p OrganizationPermission) bool {
+	t := c.token.Token().Type
+	switch t {
+	case authtoken.TypeUser:
+		role, err := c.composeOrgPermissions(ctx, orgID)
+		if err != nil {
+			// TODO: log error
+			return false
+		}
+		switch p {
+		case ReadOrg:
+			return role.ReadOrganization
+		case ManageOrg:
+			return role.ManageOrganization
+		case ReadProjects:
+			return role.ReadProjects
+		case CreateProjects:
+			return role.CreateProjects
+		case ManageProjects:
+			return role.ManageProjects
+		case ReadOrgMembers:
+			return role.ReadOrgMembers
+		case ManageOrgMembers:
+			return role.ManageOrgMembers
+		default:
+			panic(fmt.Errorf("unexpected organization permission %q", p))
+		}
+	case authtoken.TypeService:
+		return false
+	default:
+		panic(fmt.Errorf("unexpected token type %q", t))
+	}
+}
+
+func (c *authTokenClaims) CanProject(ctx context.Context, projectID string, p ProjectPermission) bool {
+	t := c.token.Token().Type
+	switch t {
+	case authtoken.TypeUser:
+		role, err := c.composeProjectPermissions(ctx, projectID)
+		if err != nil {
+			// TODO: log error
+			return false
+		}
+		switch p {
+		case ReadProject:
+			return role.ReadProject
+		case ManageProject:
+			return role.ManageProject
+		case ReadProdBranch:
+			return role.ReadProdBranch
+		case ManageProdBranch:
+			return role.ManageProdBranch
+		case ReadDevBranches:
+			return role.ReadDevBranches
+		case ManageDevBranches:
+			return role.ManageDevBranches
+		case ReadProjectMembers:
+			return role.ReadProjectMembers
+		case ManageProjectMembers:
+			return role.ManageProjectMembers
+		default:
+			// TODO: log error
+			return false
+		}
+	default:
+		// TODO: log error
+		return false
+	}
+}
+
+func (c *authTokenClaims) composeOrgPermissions(ctx context.Context, orgID string) (*database.OrganizationRole, error) {
+	var composite *database.OrganizationRole
+	directRole, err := c.admin.DB.ResolveUserOrganizationRole(ctx, c.token.OwnerID(), orgID)
+	if err != nil {
+		if !errors.Is(err, database.ErrNotFound) {
+			return nil, err
+		}
+		// no direct role, check group roles
+	}
+	if directRole != nil {
+		composite = unionOrgRoles(composite, directRole)
+	}
+	groupRoles, err := c.admin.DB.ResolveUserGroupOrgRoles(ctx, c.token.OwnerID(), orgID)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, role := range groupRoles {
+		composite = unionOrgRoles(composite, role)
+	}
+	return composite, nil
+}
```

Comment:
We should either do this in one SQL query or do the two lookups in goroutines to parallelize them. I think a unified SQL query for resolving might be simplest, what do you think?

Same applies for project roles